File: ycs-golang/utils/iupdate_decoder.go
package utils

import (
	"io"
)

// IDSDecoder interface represents a decoder for DS (Delete Set)
type IDSDecoder interface {
	Reader() io.Reader
	ResetDsCurVal()
	ReadDsClock() int64
	ReadDsLength() int64
	Close()
}

// IUpdateDecoder interface represents an update decoder
type IUpdateDecoder interface {
	IDSDecoder
	ReadLeftId() *ID
	ReadRightId() *ID
	ReadClient() int64
	ReadInfo() byte
	ReadString() string
	ReadParentInfo() bool
	ReadTypeRef() uint32
	ReadLength() int
	ReadAny() interface{}
	ReadBuffer() []byte
	ReadKey() string
	ReadJson() interface{}
}

File: src/Ycs/Structs/IUpdateDecoder.cs
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

using System;
using System.IO;

namespace Ycs.Structs
{
    public interface IDSDecoder : IDisposable
    {
        Stream Reader { get; }

        void ResetDsCurVal();
        long ReadDsClock();
        long ReadDsLength();
    }

    public interface IUpdateDecoder : IDSDecoder
    {
        ID ReadLeftId();
        ID ReadRightId();
        long ReadClient();
        byte ReadInfo();
        string ReadString();
        bool ReadParentInfo();
        uint ReadTypeRef();
        int ReadLength();
        object ReadAny();
        byte[] ReadBuffer();
        string ReadKey();
        object ReadJson();
    }
}
